File: 23.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode result = new ListNode(0);
        ListNode result_end = result;
        boolean found = false;
        do {
            int min_index = 0;
            int min = 0;
            found = false;
            for (int i=0; i<lists.length; ++i) {
                ListNode node = lists[i];
                if (node != null) {
                    if (found && node.val < min) {
                        min_index = i;
                        min = node.val;
                    }
                    else if (!found) {
                        min_index = i;
                        min = node.val;
                        found = true;
                    }
                }
            }

            if (found) {
                result_end.next = lists[min_index];
                result_end = result_end.next;
                lists[min_index] = lists[min_index].next;
            } else {
                break;
            }
        } while (found);

        return result.next;
    }
}

File: 23.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) == 0:
            return None

        res = lists[0]
        for i in range(1, len(lists)):
            res = self.merge_2_lists(res, lists[i])

        return res

    def merge_2_lists(self, list1, list2):
        dummy = ListNode()
        new_tail = dummy

        p1 = list1
        p2 = list2

        while p1 and p2:
            if p1.val <= p2.val:
                new_tail.next = p1
                new_tail = new_tail.next
                p1 = p1.next
            else:
                new_tail.next = p2
                new_tail = new_tail.next
                p2 = p2.next

        if not p1:
            new_tail.next = p2
        else:
            new_tail.next = p1

        return dummy.next
